### ISSUE TITLE
Add BONE_TRANSFORM built-in to vertex shader

### ISSUE THREAD

**Issue by pineapplemachine** [godotengine/godot-proposals#9735]
### Describe the project you are working on

I am working on a 3D game using an inverted hull shader to help character models stand out from their backgrounds. In my case, I cannot just extrude the outline hull along normal vectors, because my models do not all have smoothed normals. I have written a Blender script that computes and bakes outline extrude vectors into extra UV layers, and I have written a shader in Godot that reads that data and uses it to create the outline hull.

### Describe the problem or limitation you are having in your project

Generally, this approach is working well for me. The problem is that when the mesh this outline is applied to is animated using a skeleton, I do not have any way to apply a transform from skinning to the extrusion vectors (i.e. the same transform as Godot has automatically applied to the vertex `NORMAL` upon entering the vertex shader). Though I have access to `BONE_INDICES` and `BONE_WEIGHTS` in the vertex shader, I do not have access to the buffer containing the bone_transforms.

Related: https://github.com/godotengine/godot/blob/d48be8c5ef3d3d1eccf63a198338ded14e301e23/servers/rendering/renderer_rd/shaders/skeleton.glsl#L37

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

In my particular case, exposing `bone_transforms` data to the vertex shader - or, more conveniently and probably more realistically, a single `BONE_TRANSFORM` mat4 - will allow applying the appropriate transforms to the outline extrusion vectors in my outline shader. This would presumably be useful for many other materials as well.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

For example:

```
void vertex() {
    vec3 extrusion_vec = normalize((vec4(baked_extrusion_vec, 0.0) * BONE_TRANSFORM).xyz);
}
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I may be able to get a reasonable result by representing my baked extrusion vectors as transforms upon the normal? That's what I'm going to try next. But this is janky and inconvenient, is likely to be affected by floating point error, and at least on my end my problem could be easily solved by having access to this transform that Godot already computes per-vertex, in order to transform vertex positions, normals, and tangents, and currently is just not exposing.

### Is there a reason why this should be core and not an add-on in the asset library?

As far as I know, this is not possible to achieve with an addon.

**Comment by pineapplemachine** [comment]
> I may be able to get a reasonable result by representing my baked extrusion vectors as transforms upon the normal? That's what I'm going to try next. 

Ha, it took me a while to realize why that idea was not going to work.

I did manage to work around this issue for my own case by creating a second mesh containing the data the shader needs for the inverted hull outline, and baking the extrusion vectors into that second mesh's normals. That way, they get automatically transformed according to the pose. It's kind of messy and a pain compared to being able to bake everything into one mesh, which I could do if a `BONE_TRANSFORM` matrix were available, but at least I've still got a decent visual result.

**Comment by sjkillen** [comment]
I'm working on a vertex animation system and would also find a BONE_TRANSFORM built-in very useful.
Currently animations get baked as images that include the bone transforms for one animation.
As a result, the animations don't work with subtle changes to bone positions (such as through IK)

**Comment by TokageItLab** [comment]
1. Need to get the WEIGHT as well as TRANSFORM
2. In Godot, a Bone has two Transforms, Rest and Pose, then those should:
- BONE_WEIGHT[i]
- BONE_REST_TRANSFORM[i]
- BONE_POSE_TRANSFORM[i]
3. Since a vertex may be bound to multiple bones, they are Arrays, but the order of their indices is determined by the importer and cannot be arbitrarily determined

**Comment by pineapplemachine** [comment]
@TokageItLab I agree that it would be helpful to have access to these, too. But I think having access to the final, aggregated transformation matrix just so that the same transformation can be applied to other vectors as is already applied automatically to vertex positions, normals, and tangents would be the most valuable out of these and an important first step. I don't think it would be strictly necessary to add all of these at once.

Specifically, I'm referring to the transformation matrix defined and calculated here:  
https://github.com/godotengine/godot/blob/d48be8c5ef3d3d1eccf63a198338ded14e301e23/servers/rendering/renderer_rd/shaders/skeleton.glsl#L255

![Image](https://github.com/user-attachments/assets/7908f938-9d91-476d-9c18-28b2b5084503)

Which is then used to transform positions, normals, and tangents here:  
https://github.com/godotengine/godot/blob/d48be8c5ef3d3d1eccf63a198338ded14e301e23/servers/rendering/renderer_rd/shaders/skeleton.glsl#L282

![Image](https://github.com/user-attachments/assets/29f9768b-108b-4a72-90e3-1b5a9e54af68)

**Comment by sievaxx** [comment]
Is this actually even possible in this way? I took a look into this since I felt i could benefit from it, but wouldn't doing it require passing all bones and weights and computing the transformation in the vertex shader, defeating the point of the compute skinning, or making a giant output from the compute shader that contains every transformation matrix for every vertex to then read in the vertex shader? I couldn't think of a way to pass the data out in a nice way besides those two ideas, and neither sound nice. I think something like TokageItLabs suggestion makes more sense, as it sounds like the important thing is being able to get the current transform of a specific bone or a specific influencing bone, to then use for your own points inside the shader. However, it doesn't look like the shader system supports having arrays as "built-ins", so im not sure if it would be like that exactly and not more like `BONE_POSE_XFORM0 - BONE_POSE_XFORM7`

im trying to do source-engine styled eyeballs, which use weird object-space math at a specific 3d point, with the eye shader being more of a "window" to a sphere that then gets projected onto the triangles. the 3d positions of the "eye" are sent in global space thru the CPU to the shader instance uniforms. I'd like to avoid having to have a bit of cpu code that takes the global pose of the head bone and applies a local offset, as defining just the local offset in the shader should be possible if I had access to all the bone transforms (or easier, if I knew the "eye" window vertices were ONLY influenced by one bone, i can access the 0th index of "BONE_POSE_TRANSFORMS" outright)

**Comment by SomeGuyOnThisPlatform** [comment]
I really need this right now as well. I'm trying to make my own decal implementation, currently adding support for animated meshes, and I'd want the decals to  deform properly with the mesh rather than project onto it's deformed state.

For that I need to be able to access the rest pose of the mesh before animation is applied so that I can project the UV onto the mesh before animation and so that it can be deformed with the animation allowing decals to deform with the mesh.

At least adding something like `VERTEX_REST` that returns the same thing as VERTEX except without animation applied would be very useful, tho having `BONE_TRANSFORM` could likely be helpful here aswell.

**Comment by TokageItLab** [comment]
As I mentioned in https://github.com/godotengine/godot/pull/110120, I consider this necessary for Gizmo objects where we might modify rotation such as the case only want to bind the bone position. It should be possible to cancel rotation within the shader by exposing skin/joint matrix parameters, without modifying the skinning system itself.
